File: BangHash/sample_interpreter.cc
// Includes
#include "config.hh"
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
//#include <malloc.h>
#include <errno.h>
#include <libgen.h>

#if MAIN_VARIATION == 1 || MAIN_VARIATION == 2
#include "env2lib.hh"
#endif
#include "sample_interpreter.hh"

// #############################################################################

// * load_script *****************************************************
argv_t load_script(char *scriptname) {
  char *lineptr = NULL;
  size_t linesze = 0;
  int read;
  int i=0;
  int max_script_lines = 1024;
  argv_t sc;
  sc.argv = (char **)calloc(max_script_lines, sizeof(char *));

  // open files
  FILE *script = fopen(scriptname, "r");
  if(script == NULL) {
    fprintf(stderr, "%s error: file open failed for '%s'\n", Argv0, scriptname);
    exit(1);
  }

  // read file
  while((read = getline(&lineptr, &linesze, script)) != -1) {
    sc.argv[i++] = strdup(lineptr);
    if(i==max_script_lines-2) {
      fprintf(stderr, "%s error: script exceeded max lines '%d' tuncating\n", Argv0, max_script_lines);
      break;
    }
  }
  sc.argv[i] = NULL;
  sc.argc = i;
  free(lineptr);
  fclose(script);

  return(sc);
}

// * run_bash *****************************************************
void run_bash(argv_t ia, argv_t sa) {
  char *iargvs;
  char *sargvs;
  argv_t sc;
  int i;

  iargvs = join_array(ia.argv, ia.argc);
  sargvs = join_array(sa.argv+1, sa.argc-1);

  sc = load_script(sa.argv[0]);
  // open files
#if 1
  FILE *shell  = popen("/bin/bash --norc --noprofile -s", "w");
#else
  FILE *shell  = popen("/bin/bash --norc --noprofile -i", "w");
#endif
  if(shell == NULL) {
    fprintf(stderr, "%s error: pipe open failed for '%s'\n", Argv0, "bash");
    exit(1);
  }


  // set vars
  fprintf(shell, "typeset -a argv=(%s)\n", iargvs);
  fprintf(shell, "BASH_ARGV0=%s\n", sa.argv[0]);  // bash v5 feature
  fprintf(shell, "set -- %s\n", sargvs);

  // execute script
  for (i=0;i<sc.argc;i++) {
    fprintf(shell, "%s\n", sc.argv[i]);
  }
  pclose(shell);
}

// #############################################################################

#if MAIN_VARIATION == 1

// *************************************************************
// * main1 - Call env2()
// ************************************************************

// Define Globals
char *Argv0;                   // Name of program
int   Debug;                   // Debug flag
hash_t flags;                  // Place for flags (xxx move to private var)

using namespace std;

// * usage *****************************************************
void usage(int ret) {
  printf("Usage\n");
  exit(ret);
}

// * main1 *****************************************************
int main(int argc, char **argv) {
  /// PARSE ARAGS //////////////////////////////////////////////////////////////
  // define local vars
  int i, scr_loc;
  argv_t o, e;
  
  o.argc = argc;
  o.argv = argv;

  // set global vars
  Argv0 = o.argv[0];
  Debug = 1;
  setvbuf(stdout, NULL, _IONBF, 0);

  // make sure we have at least one arg
  if(o.argc < 2) {
    usage(1);
  }

  // set flags (S=split_string E=env2 V=vars2)
  init_flags();
  flags["delim"]  = string("~~");  // E search

  // Call the main env2() function
  try {
    e = env2(o);
  } catch (StdException &exc) {
    fprintf(stderr, "%s error: %s\n", Argv0, exc.what());
    usage(1);
  } catch (...) {
    fprintf(stderr, "%s error: caught unknown exception\n", Argv0);
    exit(1);
  }

  /// DUMP ARAGS //////////////////////////////////////////////////////////////
  printf("argv[0]='%s'\n",o.argv[0]);
  for(i=0; i<e.argc; i++) {
    printf("argv[%i]='%s'\n",i+1,e.argv[i]);
    if((i>0) && (e.argv[i][0] != '-')) scr_loc=i;
  }

  /// PARSE SCRIPT /////////////////////////////////////////////////////////////
  // fix up iargv and sargv
  argv_t ia;
  ia.argv    =  (char **)calloc(e.argc, sizeof(char *));
  ia.argv[0] = o.argv[0];
  for(i=1;i<scr_loc+1;i++) ia.argv[i] = e.argv[i-1];
  ia.argv[i] = NULL;
  ia.argc    = scr_loc+1;

  argv_t sa;
  sa.argv = e.argv+= scr_loc;
  sa.argc = e.argc-scr_loc;
  // call run_bash
  run_bash(ia, sa);

  return(0);
}

// #############################################################################

#elif MAIN_VARIATION == 2


// *************************************************************
// * main2 - Call split_and_merge()
// ************************************************************


// Define Globals
char *Argv0;                   // Name of program
int   Debug;                   // Debug flag
hash_t flags;                  // Place for flags (xxx move to private var)

using namespace std;

// * usage *****************************************************
void usage(int ret) {
  printf("Usage\n");
  exit(ret);
}

// * main2 *****************************************************
int main(int argc, char **argv) {
  /// PARSE ARAGS //////////////////////////////////////////////////////////////
  argv_t o, e;
  o.argc = argc;
  o.argv = argv;
  int i = 0;

  // set global vars
  Argv0 = o.argv[0];
  Debug = 1;
  setvbuf(stdout, NULL, _IONBF, 0);

  // make sure we have the right number of args
  if(o.argc == 1) {
    usage(1);
  }

  // set flags (S=split_string E=env2 V=vars2)
  init_flags();

  try {
#if KERNEL_SPLIT == 1
  o = read_hashbang(o);    // get hashbang line from script
#endif
#if KERNEL_SPLIT != 2
  e = split_and_merge(o);  // split up argv[1]
#else
  e = o;
#endif
  } catch (StdException &exc) {
    fprintf(stderr, "%s error: %s\n", Argv0, exc.what());
    usage(1);
  } catch (...) {
    fprintf(stderr, "%s error: caught unknown exception\n", Argv0);
    exit(1);
  }

  /// DUMP ARAGS //////////////////////////////////////////////////////////////
  for(i=0; i<e.argc; i++) {
    printf("argv[%i]='%s'\n",i,e.argv[i]);
  }

  /// PARSE SCRIPT //////////////////////////////////////////////////////////////
  //run_bash(ia, sa);

  return(0);
}

// #############################################################################

#else
  
#error "Must define MAIN_VARIATION as 1 or 2"

#endif

File: BangHash/sample_interpreter.pl
#!/usr/bin/perl -w
#
# This interpreter is silly and should be deleted.
#      

use strict;
use warnings;

sub usage {
  print "Usage: $0 [<iflags>] <script> [<sflags>] [<sargs>]\n";
  exit(1);
}

sub parse_flags {
  my($iargs, $sargs) = @_;
  my @sargs_t;
  my %flags;
  my $farg=undef;
  my $delim=0;
  my $oc;
  my $ia;
  
  # NOTE: This does not handle quoted or escaped whitespaces
  $ia = shift @$iargs;
A:for my $a (@$iargs) {
B:  for my $b (split /\s+/, $a) {
      if($b eq "~~") {
        $delim=1;
      } elsif ($delim) {
        push(@sargs_t, $b);
      } elsif ($b =~ s/^-//) {
        $farg= undef;
        $oc  = undef;
C:      for my $c (split //, $b) {
          if($c =~ /[:=]/) {
            $farg=$oc;
            $flags{$farg} = "";
          } else {
            if(defined $farg) {
              $flags{$farg} .= $c;
            } else {
              $flags{$c} = 1;
            }
          }
        } continue {
          $oc = $c;
        }
      } else {
        last A;
      }
    }
  }
  unshift @$iargs, $ia;
  splice(@$sargs, 1, 0, @sargs_t);
  # print found flags #####
  print "---my flags---\n";
  for my $f (sort keys %flags) {
    print "$f='$flags{$f}'\n";
  }
}

sub print_args {
  # vars
  my @pargs; # Perl args        (super interpreter)
  my @iargs; # Interpreter agrs (this program)
  my @sargs; # Script args      (child to be run)
  my $i;
  my $found;
  my $fh;
  my $line;
  local $" = " ";
  # get perl args ############################
  push(@pargs, $^X);
  if(-d "/proc" and -f "/proc/self/cmdline") {
    open($fh, "<", "/proc/self/cmdline");
    $line = <$fh>;
    close($fh);
    $line =~ s/\c@/\cj/g;
  } else {
    $line = (`ps -p $$ -o command=`)[0];
    $line =~ s/ /\cj/g;
  }
  $i = 1;
  $found=0;
  for my $a (split /\cj/, $line) {
    if($a =~ /^-/) {
      if($found>1) {
        1;
      } else {
        push(@pargs, $a);
      } 
    } else {
      $found++;
    }
  }
  # get inter args ###########################
  push(@iargs, $0);
  $i = 1;
  $found=0;
  for my $a (@ARGV) {
    if($a =~ /^[-~]/) {
      if($found) {
        push(@sargs, $a);
      } else {
        push(@iargs, $a);
      } 
    } else {
      $found++;
      push(@sargs, $a);
    }
  }
  # print #####
  print "---perl args---\n";
  for $i (0..$#pargs) { print "argv[$i]=$pargs[$i]\n"; }
  print "---inter args---\n";
  for $i (0..$#iargs) { print "argv[$i]=$iargs[$i]\n"; }
  print "---script args---\n";
  for $i (0..$#sargs) { print "argv[$i]=$sargs[$i]\n"; }

  return(\@iargs, \@sargs);
}

sub call_bash {
  my($iargs, $sargs) = @_;
  my $fh;
  my @lines;
  my $sa;
  local $" = "' '";

  # exists?  
  if(! -f $$sargs[0]) {
    print "error: script not found: $$sargs[0]\n";
    return(1);
  }
  # read in
  open($fh, "<", $$sargs[0]);
  @lines = <$fh>;
  close($fh);
  # exec
  open($fh, "|-", "/bin/bash --norc --noprofile -s");
  # set vars
  print $fh "typeset -a argv=('@$iargs')\n";
  print $fh "BASH_ARGV0=$$sargs[0]\n";
  $sa = shift @$sargs;
  print $fh "set -- '@$sargs'\n";
  unshift @$sargs, $sa;
  # run script
  for my $line (@lines) {
    print $fh $line;
  }
  close($fh);
}

sub main {
  #### vars
  my $iargs;
  my $sargs;
  #### Usage
  if(!@ARGV) {
    usage();
  }
  #### Print my args
  ($iargs, $sargs) = print_args();
  #### Parse flags
  parse_flags($iargs, $sargs); # iargs=in sargs=in+out
  #### Execute script
  call_bash($iargs, $sargs);
}
main();
